### PR TITLE
[FIXED JENKINS-28188] - Fix for plugin compatibility when used anonymous class cause

### PR DESCRIPTION
**Pull request by lanwen**
Fixes [JENKINS-28188](https://issues.jenkins-ci.org/browse/JENKINS-28188)
Example for reproduce (test added):
```java
 FreeStyleBuild build = jenkins.createFreeStyleProject().scheduleBuild2(0, new Cause() {
            @Override
            public String getShortDescription() {
                return "This build was started by a hobbit Bilbo. Bilbo Baggins";
            }
        }).get();
```

Exception:
```
[EnvInject] - Loading node environment variables.
FATAL: String index out of range: -1
http://stacktrace.jenkins-ci.org/search?query=java.lang.StringIndexOutOfBoundsException: String index out of range: -1
at java.lang.String.substring(String.java:1875) <http://stacktrace.jenkins-ci.org/search/?query=java.lang.String.substring&entity=method>
at org.jenkinsci.plugins.envinject.service.BuildCauseRetriever.buildCauseEnvironmentVariables(BuildCauseRetriever.java:71) <http://stacktrace.jenkins-ci.org/search/?query=org.jenkinsci.plugins.envinject.service.BuildCauseRetriever.buildCauseEnvironmentVariables&entity=method>
at org.jenkinsci.plugins.envinject.service.BuildCauseRetriever.getTriggeredCause(BuildCauseRetriever.java:40) <http://stacktrace.jenkins-ci.org/search/?query=org.jenkinsci.plugins.envinject.service.BuildCauseRetriever.getTriggeredCause&entity=method>
at org.jenkinsci.plugins.envinject.service.EnvInjectVariableGetter.getBuildVariables(EnvInjectVariableGetter.java:111) <http://stacktrace.jenkins-ci.org/search/?query=org.jenkinsci.plugins.envinject.service.EnvInjectVariableGetter.getBuildVariables&entity=method>
at org.jenkinsci.plugins.envinject.EnvInjectListener.setUpEnvironmentWithoutJobPropertyObject(EnvInjectListener.java:231) <http://stacktrace.jenkins-ci.org/search/?query=org.jenkinsci.plugins.envinject.EnvInjectListener.setUpEnvironmentWithoutJobPropertyObject&entity=method>
at org.jenkinsci.plugins.envinject.EnvInjectListener.setUpEnvironment(EnvInjectListener.java:46) <http://stacktrace.jenkins-ci.org/search/?query=org.jenkinsci.plugins.envinject.EnvInjectListener.setUpEnvironment&entity=method>
at hudson.model.AbstractBuild$AbstractBuildExecution.createLauncher(AbstractBuild.java:587) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.AbstractBuild$AbstractBuildExecution.createLauncher&entity=method>
at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:491) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.AbstractBuild$AbstractBuildExecution.run&entity=method>
at hudson.model.Run.execute(Run.java:1759) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Run.execute&entity=method>
at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.FreeStyleBuild.run&entity=method>
at hudson.model.ResourceController.execute(ResourceController.java:89) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.ResourceController.execute&entity=method>
at hudson.model.Executor.run(Executor.java:240) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Executor.run&entity=method>
```

Also:
- Refactor test for this class to junit4 style

#### PLEASE REVIEW BY COMMITS
(not whole changeset at once)

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/envinject-plugin/45)

### CODE REVIEW

**Comment by lanwen** [comment]
Hmm, 
```
testPropertiesContentCustomWorkspace(org.jenkinsci.plugins.envinject.EnvInjectBuildWrapperTest): expected:</[tmp/hudson5040398227488372587test/customWorkspace]/materialize_workspa...> 
but was:</[scratch/jenkins/workspace/plugins/envinject-plugin]/materialize_workspa...>
```
think it not my fault

**Comment by lanwen** [comment]
seems like 
```java
FreeStyleProject project = j.createFreeStyleProject();
        project.setCustomWorkspace(customWorkspaceValue);
```
doesn't work on jenkins.ci.cloudbees.com... maybe assume this test by `GIT_PR_URL` var for example?

**Comment by lanwen** [comment]
Problem with test is that host jenkins overrides `workspace` env in test-jenkins. Ill think how to fix it

**Comment by lanwen** [comment]
so i've assume no any `GIT_PR_FROM` env vars is set

File: pom.xml [hunk 1]
```diff
@@ -40,7 +40,7 @@
         <maven.compiler.target>1.6</maven.compiler.target>
         <envinject.lib.version>1.21</envinject.lib.version>
         <ivy.plugin.version>1.21</ivy.plugin.version>
-        <junit.version>4.9</junit.version>
+        <junit.version>4.12</junit.version>
```
Comment (KostyaSha) [inline pom.xml]: 4.12 for example fails with mockito. 4.11 works fine

Comment (lanwen) [inline pom.xml]: hmm, is there any issue? It works fine in CI and on my local maven instance

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 2]
```diff
@@ -59,16 +61,15 @@ private static void insertRootCauseNames(Set<String> causeNames, Cause cause, in
...
-                all.append(",");
-                all.append(name);
+            if (isNotBlank(name)) {
+                triggerVars.put(on("_").join(envBase, name), "true");
+                nonEmptyNames.add(name);
             }
         }
         // add variable containing all the trigger names
-        triggerVars.put(envBase, all.toString().substring(1));
+        triggerVars.put(envBase, on(",").join(nonEmptyNames));
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Conflicts with #49 . Please avoid bulk changes (especially when they are not listed in the PRs summary)

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 3]
```diff
@@ -1,121 +0,0 @@
-package org.jenkinsci.plugins.envinject;
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: Seems Git has not noticed the move operation. It will cause conflicts with other PRs, so it's better to revert the change

File: src/test/java/org/jenkinsci/plugins/envinject/EnvInjectBuildWrapperTest.java [hunk 4]
```diff
@@ -1,27 +1,37 @@
...
+import static org.jenkinsci.plugins.envinject.matchers.WithEnvInjectActionMatchers.withEnvInjectAction;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assume.assumeThat;
 
 public class EnvInjectBuildWrapperTest {
 
-    public @Rule JenkinsRule j = new JenkinsRule();
+    @ClassRule
+    public static JenkinsRule j = new JenkinsRule();
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/envinject/EnvInjectBuildWrapperTest.java]: :-1: . It may cause a cross-influence between tests

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: This is not conflict, its the same fix. But with tests

File: src/test/java/org/jenkinsci/plugins/envinject/EnvInjectBuildWrapperTest.java (hunk 4 above)
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/EnvInjectBuildWrapperTest.java]: No, it can't - every test create new job which not intefere with each other. I've change it to class rule with attention to war between tests.

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java (hunk 3 above)
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: ok, will be soon.

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: all changes listed in summary

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java (hunk 3 above)
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: I've completely rewrite tests in this class. This is the main cause why git not handles this move. Tests save the logic but became simpler. And understandable - just read them.

Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: In such case you should manually notify git about moves or split move and further changes into two commits

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Agreed. A JIRA reference helps even if the initial description was too narrow-specific

**Comment by lanwen** [comment]
@oleg-nenashev Hello, i've reworked PR. Please rereview.
Removed fix of CI build and renaming of package (will be in separate PRs)

**Comment by oleg-nenashev** [comment]
Looks good to me. 

@dcoraboeuf, @KostyaSha  
Do you agree with the merge?

**Comment by dcoraboeuf** [comment]
Fine with me. Thanks.

**Comment by KostyaSha** [comment]
Have no opinion.

**Comment by lanwen** [comment]
fixed test in #51
